### PR TITLE
[RLlib] DreamerV3: Add MultiDiscrete action space support

### PR DESCRIPTION
**Pull request by merhovon**
## Description

This PR adds support for **MultiDiscrete action spaces** in the DreamerV3 algorithm (PyTorch only).

### Key Changes:

**1. ActorNetwork (`actor_network.py`)**
- Added `MultiDiscreteOneHotDistribution` class that wraps multiple `OneHotCategorical` distributions
- Extended `__init__` to create `nn.ModuleList` of MLP heads (one per sub-action dimension) with shared backbone
- Modified `forward()` to apply unimix (1% uniform) and straight-through gradients per sub-action
- Updated `get_action_dist_object()` to return the custom distribution

**2. DreamerModel (`dreamer_model.py`)**
- `get_initial_state()`: Compute action dimension as `sum(nvec)` for MultiDiscrete
- `dream_trajectory()` & `dream_trajectory_with_burn_in()`: Handle one-hot↔int conversion and random action generation for MultiDiscrete spaces

**3. RLModule (`dreamerv3_torch_rl_module.py`)**
- Extended `_forward_inference_or_exploration_helper()` to convert concatenated one-hot actions to multi-int format for environment compatibility

### Implementation Approach:
- Uses concatenated one-hot vectors for the world model (consistent with existing Discrete handling)
- Separate categorical distributions with **unimix (1% uniform mixing)** for each sub-action dimension

### Testing:
- Added unit tests in `test_multidiscrete_unit.py` covering:
- Network initialization with various MultiDiscrete configs
- Forward pass output shapes
- Distribution sampling and log probability computation

## Related issues

Closes #59608

## Additional information

- **PyTorch only**: TensorFlow support is not included as it's being deprecated
- **Tested configurations**: `MultiDiscrete([3, 4])`, `MultiDiscrete([2, 3, 4])`, `MultiDiscrete([5, 5, 5, 5])`
- Based on guidance from @pseudo-rnd-thoughts in the issue discussion

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
Thanks @merhovon I will try to have a look soon

File: rllib/algorithms/dreamerv3/torch/models/actor_network.py [hunk 1]
```diff
@@ -3,15 +3,77 @@
...
+from ray.rllib.algorithms.dreamerv3.torch.models.components import (
+    dreamerv3_normal_initializer,
+)
 from ray.rllib.algorithms.dreamerv3.torch.models.components.mlp import MLP
 from ray.rllib.utils.framework import try_import_torch
 
 torch, nn = try_import_torch()
 
 
+class MultiDiscreteOneHotDistribution:
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/dreamerv3/torch/models/actor_network.py:20]: Is this unique to Dreamer-V3, can we use one of the distributions in `rllib/core/distributional`?

File: rllib/algorithms/dreamerv3/torch/models/actor_network.py [hunk 2]
```diff
@@ -172,6 +282,15 @@ def get_action_dist_object(self, action_dist_params_T_B):
...
             distr = torch.distributions.Independent(distr, len(self.action_space.shape))
 
+        elif isinstance(self.action_space, gym.spaces.MultiDiscrete):
+            # Split logits according to actions_dim and create distributions
+            split_logits = torch.split(action_dist_params_T_B, self.actions_dim, dim=-1)
+            distributions = [
+                torch.distributions.OneHotCategorical(logits=logits)
+                for logits in split_logits
+            ]
+            distr = MultiDiscreteOneHotDistribution(distributions, self.actions_dim)
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: I think we can use TorchMultiDistribution or TorchMultiCategorical instead of a custom distribution

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Overall looks good. @merhovon Have you trained a policy with this to test that a policy can learn with it

File: rllib/algorithms/dreamerv3/torch/models/actor_network.py (hunk 2 above)
Comment (merhovon) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: The existing `TorchMultiCategorical` uses **integer representations** for actions (`sample()` returns `[B, num_cats]` of integers, `logp()` expects integers).

 DreamerV3 specifically requires **concatenated one-hot vectors** for the world model (consistent with how `Discrete` actions are handled via `torch.distributions.OneHotCategorical`), and needs a `mode` property for deterministic actions during inference.

 Key differences:
 - `sample()` → needs concatenated one-hot `[B, sum(nvec)]`, not integers `[B, num_cats]`
 - `log_prob()` → expects one-hot input, not integers  
 - `mode` property → returns one-hot mode for inference

I could potentially create a wrapper around `TorchMultiCategorical` that handles the conversion, but this would add complexity. The custom class is ~30 lines and directly mirrors `torch.distributions.OneHotCategorical` which DreamerV3 already uses for `Discrete`.

**Comment by pseudo-rnd-thoughts** [comment]
@merhovon Run pre-commit to resolve most of the microcheck issues

File: rllib/algorithms/dreamerv3/torch/models/actor_network.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: Could you reference where in the paper or the official code this one-hot vectors for integer representations is used? I thought the distribution is for the output actions not the input representation

Comment (merhovon) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: **Evidence from Danijar Hafner's official implementation:**
 
 In [`rssm.py`](https://github.com/danijar/dreamerv3/blob/main/dreamerv3/rssm.py), the `_observe` and `imagine` methods use:
 ```python
 action = nn.DictConcat(self.act_space, 1)(action)
 ```
 This concatenates all actions into a single vector. For discrete actions, `OneHotCategorical` distributions are used, producing one-hot vectors.
 
 The action is then fed directly into the dynamics core:
 ```python
 def _core(self, deter, stoch, action):
     ...
     x2 = self.sub('dynin2', nn.Linear, self.hidden, **self.kw)(action)
 ```
 
 **Evidence from existing RLlib DreamerV3 code:**
 
 - [`sequence_model.py` L34-35](https://github.com/ray-project/ray/blob/master/rllib/algorithms/dreamerv3/torch/models/components/sequence_model.py#L34-L35): *"Initial `a` is the zero 'one hot' action, e.g. [0.0, 0.0] for Discrete(2)."*
 - [`dreamerv3.py` L605-619](https://github.com/ray-project/ray/blob/master/rllib/algorithms/dreamerv3/dreamerv3.py#L605-L619): Integer actions are converted to one-hot **before** training the world model.
 
 **Why `TorchMultiCategorical` doesn't work:**
 
 RLlib's `TorchMultiCategorical` returns integer samples with shape `[B, num_categories]`. But DreamerV3 needs **concatenated one-hot vectors** with shape `[B, sum(nvec)]` for the world model input. This is why I created `MultiDiscreteOneHotDistribution` - it follows the same pattern as the existing `Discrete` support which uses `torch.distributions.OneHotCategorical`.

Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: Re-reading your first message, it said that the distribution would be needed for the world model rather than the actor network. Did I understand that correctly? 
For the actor network, why can't we use TorchMultiCategorical which is used through the rest of the project for MultiDiscrete support?

Comment (merhovon) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: The key insight is that DreamerV3's existing `Discrete` implementation **doesn't use RLlib's `TorchCategorical`** - it uses `torch.distributions.OneHotCategorical` directly and has its own specialized log_prob computation in the learner.
 
 Looking at dreamerv3_torch_learner.py lines 569-580:
 ```python
 # For Discrete - manual logp via dot product:
 logp_actions_dreamed = torch.sum(actions_dreamed * dist_params, dim=-1)
 ```
 
 This is because:
 1. DreamerV3 applies "unimix" (99% NN + 1% uniform) to create mixed probabilities
 2. The `distr_params` are already `log(probs)`, not raw logits
 3. The logp is computed manually as `sum(one_hot_action * log_probs)`
 
 So `TorchMultiCategorical.logp()` (which expects integer inputs) wouldn't directly work with DreamerV3's existing architecture.
 
 **Two options:**
 1. **Current approach**: Follow the same pattern as `Discrete` - use `OneHotCategorical` per sub-action and wrap in `MultiDiscreteOneHotDistribution`. This maintains consistency with existing DreamerV3 code.
 
 2. **Alternative**: Refactor to use `TorchMultiCategorical`, but this would require:
    - Converting one-hot → integers before `logp()`
    - Converting integers → one-hot after `sample()`
    - Potentially modifying the learner's logp computation
    - Changing how actions are stored/processed throughout the dream loop

Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py:292]: OK that makes sense, I think this make sense. Thanks for explaining.

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py [hunk 3]
```diff
@@ -440,6 +448,15 @@ def dream_trajectory_with_burn_in(
...
                     a = torch.nn.functional.one_hot(a, num_classes=self.action_space.n)
+                elif isinstance(self.action_space, gym.spaces.MultiDiscrete):
+                    # Random one-hot per sub-action, then concatenate
+                    one_hots = []
+                    for n in self.action_space.nvec:
+                        rand_idx = torch.randint(int(n), (B,), dtype=torch.int64)
+                        one_hots.append(
+                            torch.nn.functional.one_hot(rand_idx, num_classes=int(n))
+                        )
+                    a = torch.cat(one_hots, dim=-1).float()
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dreamerv3/torch/models/dreamer_model.py]: @merhovon Could you fix this

File: rllib/algorithms/dreamerv3/torch/models/actor_network.py [hunk 4]
```diff
@@ -12,6 +14,63 @@
...
+    @property
+    def mode(self):
+        """Return concatenated one-hot modes (argmax) for all sub-actions."""
+        modes = []
+        for dist in self.distributions:
+            probs = dist.probs
+            one_hot = torch.zeros_like(probs)
+            one_hot.scatter_(-1, probs.argmax(dim=-1, keepdim=True), 1.0)
+            modes.append(one_hot)
+        return torch.cat(modes, dim=-1)
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply rllib/algorithms/dreamerv3/torch/models/actor_network.py]: @merhovon Could you fix this

**Comment by pseudo-rnd-thoughts** [comment]
@merhovon STOP merging with master. You need to run pre-commit.
Also, have you trained an agent with this? Could you provide the training curve for it

**Comment by merhovon** [comment]
> @merhovon STOP merging with master. You need to run pre-commit. Also, have you trained an agent with this? Could you provide the training curve for it

Pre-commit formatting has been applied.
Regarding the training curve: I'm currently on a business trip and will provide training results once I'm back.

**Comment by merhovon** [comment]
@pseudo-rnd-thoughts 

I ran a 100k environment steps training run on `KungFuMaster-v5` using the `MultiDiscrete([2, 7])` action space (Discrete(14) decomposed via mixed-radix encoding).

**Setup:**
- DreamerV3 model size: `XS`
- `training_ratio=64`, `batch_size_B=16`
- `num_envs_per_env_runner=32`
- 1x GPU (RTX 2070), local learner
- Atari wrappers: 64x64 RGB, normalized

**Results (100k env steps, 433 episodes):**

image

!training_curve

| Metric | Value |
|---|---|
| Max Episode Return | 800 |
| Peak Rolling Avg (window=20) | 210 |
| Episodes with Reward > 0 | 77 / 398 (19%) |

**Observations:**
The agent shows clear learning signal in the first ~50k steps, with rewards climbing from 0 to 800. In the second half, rewards mostly drop back to 0 with occasional spikes (200, 600). This is expected behavior given the low `training_ratio=64` — the DreamerV3 paper recommends 512–1024 for Atari, meaning the world model doesn't get enough gradient updates relative to collected experience at this ratio. With a proper training ratio the learning curve would be more stable, but this would require significantly longer wall-clock time.

**What this demonstrates:**
1. DreamerV3 initializes and trains without errors using `MultiDiscrete([2, 7])`
2. The full pipeline works end-to-end: MultiDiscrete wrapper → one-hot encoding → per-sub-action actor heads → int reconstruction
3. The agent learns meaningful behavior (non-random rewards up to 800)

Is this sufficient as a proof-of-concept for the PR, or would you prefer a longer run with `training_ratio=512`?
